### PR TITLE
IMG-244 - Add support for parsing CMETAA TREs

### PR DESCRIPTION
**Pull request by derekwilhelm**
Related JIRA ticket: [IMG-244](https://codice.atlassian.net/browse/IMG-244)

### CODE REVIEW

File: core/src/test/java/org/codice/imaging/nitf/core/tre/impl/CMETAA_Test.java [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+        assertEquals("zzBJKeeeeeeeee17900nnnn", cmetaa.getFieldValue("CG_MGRSCORLL"));
+        assertEquals("1800.00", cmetaa.getFieldValue("CG_MGRS_CONFIDENCE"));
+        assertEquals("           ", cmetaa.getFieldValue("CG_MGRS_PAD"));
+    }
+
+    @Test
+    public void testCmetaaNa() throws NitfFormatException {
+        Tre cmetaa = readTre("/CMETAA/CMETAA_NA.txt");
+        assertNotNull(cmetaa);
+        //System.out.println("NA Raw Data: " + new String(cmetaa.getRawData()));
```
Comment (bradh) [inline core/src/test/java/org/codice/imaging/nitf/core/tre/impl/CMETAA_Test.java]: Can you please remove this line?

File: core/src/test/java/org/codice/imaging/nitf/core/tre/impl/CMETAA_Test.java [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        assertEquals("1.7300000", cmetaa.getFieldValue("CG_PATCH_LAT_CONFIDENCE"));
+        assertEquals("1.7400000", cmetaa.getFieldValue("CG_PATCH_LONG_CONFIDENCE"));
+    }
+
+    @Test
+    public void testCmetaaMgrs() throws NitfFormatException {
+        Tre cmetaa = readTre("/CMETAA/CMETAA_MGRS.txt");
+
+        assertNotNull(cmetaa);
+        //System.out.println("MGRS Raw Data: " + new String(cmetaa.getRawData()));
```
Comment (bradh) [inline core/src/test/java/org/codice/imaging/nitf/core/tre/impl/CMETAA_Test.java]: Can you please remove this line?

File: core/src/test/java/org/codice/imaging/nitf/core/tre/impl/CMETAA_Test.java [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+import org.codice.imaging.nitf.core.tre.TreSource;
+import org.junit.Test;
+
+public class CMETAA_Test {
+
+    @Test
+    public void testCmetaaGeod() throws NitfFormatException {
+        Tre cmetaa = readTre("/CMETAA/CMETAA_GEOD.txt");
+        assertNotNull(cmetaa);
+        //System.out.println("GEOD Raw Data: " + new String(cmetaa.getRawData()));
```
Comment (bradh) [inline core/src/test/java/org/codice/imaging/nitf/core/tre/impl/CMETAA_Test.java]: Can you please remove this line?

**Comment by derekwilhelm** [comment]
Yes, this has been tested with production examples.

Yes, I would be good with this getting upstreamed to GDAL/OGR.

File: core/src/main/resources/nitf_spec.xml [hunk 4]
```diff
@@ -199,6 +199,214 @@
...
+        <field name="CG_OPM" length="7" type="real" minval="0.0" maxval="999.999" unit="miliarcsec"/>
+        <field name="CG_MODEL" length="5" type="string"/>
+        <field name="CG_AMPT_X" length="13" type="real" minval="-99999999.999" maxval="99999999.999"/>
+        <field name="CG_AMPT_Y" length="13" type="real" minval="-99999999.999" maxval="99999999.999"/>
+        <field name="CG_AMPT_Z" length="13" type="real" minval="-99999999.999" maxval="99999999.999"/>
+        <field name="CG_AP_CONF_XY" length="6" type="real" minval="0.0" maxval="999.99" unit="meters"/>
+        <field name="CG_AP_CONF_Z" length="6" type="real" minval="0.0" maxval="999.99" unit="meters"/>
+        <field name="CG_APCEN_X" length="13" type="real" minval="-99999999.999" maxval="99999999.99"/>
+        <field name="CG_APCEN_Y" length="13" type="real" minval="-99999999.999" maxval="99999999.99"/>
+        <field name="CG_APCEN_Z" length="13" type="real" minval="-99999999.999" maxval="99999999.99"/>
```
Comment (bradh) [inline core/src/main/resources/nitf_spec.xml]: I think you could set unit to be metres (meters) here.

File: core/src/main/resources/nitf_spec.xml [hunk 5]
```diff
@@ -199,6 +199,214 @@
...
+        <field name="CG_EAST" length="8" type="real" minval="0.0" maxval="359.9999" unit="degrees"/>
+        <field name="CG_RLOS" length="8" type="real" minval="0.0" maxval="359.9999" unit="degrees"/>
+        <field name="CG_LOS_CONFIDENCE" length="6" type="real" minval="0.0" maxval="9.9999" unit="degrees"/>
+        <field name="CG_LAYOVER" length="8" type="real" minval="0.0" maxval="359.9999" unit="degrees"/>
+        <field name="CG_SHADOW" length="8" type="real" minval="0.0" maxval="359.9999" unit="degrees"/>
+        <field name="CG_OPM" length="7" type="real" minval="0.0" maxval="999.999" unit="miliarcsec"/>
+        <field name="CG_MODEL" length="5" type="string"/>
+        <field name="CG_AMPT_X" length="13" type="real" minval="-99999999.999" maxval="99999999.999"/>
+        <field name="CG_AMPT_Y" length="13" type="real" minval="-99999999.999" maxval="99999999.999"/>
+        <field name="CG_AMPT_Z" length="13" type="real" minval="-99999999.999" maxval="99999999.999"/>
```
Comment (bradh) [inline core/src/main/resources/nitf_spec.xml]: I think you could set unit to be metres (meters) here.

File: core/src/main/resources/nitf_spec.xml [hunk 6]
```diff
@@ -199,6 +199,214 @@
...
+        <field name="CG_APCEN_Z" length="13" type="real" minval="-99999999.999" maxval="99999999.99"/>
+        <field name="CG_APER_CONF_XY" length="6" type="real" minval="0.0" maxval="999.99" unit="meters"/>
+        <field name="CG_APER_CONF_Z" length="6" type="real" minval="0.0" maxval="999.99" unit="meters"/>
+        <field name="CG_FPNUV_X" length="9" type="real" minval="-1.0" maxval="1.0"/>
+        <field name="CG_FPNUV_Y" length="9" type="real" minval="-1.0" maxval="1.0"/>
+        <field name="CG_FPNUV_Z" length="9" type="real" minval="-1.0" maxval="1.0"/>
+        <field name="CG_IDPNUVX" length="9" type="real" minval="-1.0" maxval="1.0"/>
+        <field name="CG_IDPNUVY" length="9" type="real" minval="-1.0" maxval="1.0"/>
+        <field name="CG_IDPNUVZ" length="9" type="real" minval="-1.0" maxval="1.0"/>
+        <field name="CG_SCECN_X" length="13" type="real" minval="-99999999.999" maxval="99999999.999" unit="meters"/>
```
Comment (bradh) [inline core/src/main/resources/nitf_spec.xml]: The units can be degrees or metres here (depends on `CG_MODEL` value). Suggest just removing it for the X and Y cases.

File: core/src/main/resources/nitf_spec.xml [hunk 7]
```diff
@@ -199,6 +199,214 @@
...
+        <field name="CG_SCECN_Y" length="13" type="real" minval="-99999999.999" maxval="99999999.999" unit="meters"/>
```
Comment (bradh) [inline core/src/main/resources/nitf_spec.xml]: The units can be degrees or metres here (depends on `CG_MODEL` value). Suggest just removing it for the X and Y cases.

File: core/src/main/resources/nitf_spec.xml [hunk 8]
```diff
@@ -199,6 +199,214 @@
...
+            <field name="CG_PATCH_LNGCEN" length="12" type="real" minval="-179.9999999" maxval="179.9999999" unit="degrees"/>
+            <field name="CG_PATCH_LTCORUL" length="11" type="real" minval="-89.9999999" maxval="89.9999999" unit="degrees"/>
+            <field name="CG_PATCH_LGCORUL" length="12" type="real" minval="-179.9999999" maxval="179.9999999" unit="degrees"/>
+            <field name="CG_PATCH_LTCORUR" length="11" type="real" minval="-89.9999999" maxval="89.9999999" unit="degrees"/>
+            <field name="CG_PATCH_LGCORUR" length="12" type="real" minval="-179.9999999" maxval="179.9999999" unit="degrees"/>
+            <field name="CG_PATCH_LTCORLR" length="11" type="real" minval="-89.9999999" maxval="89.9999999" unit="degrees"/>
+            <field name="CG_PATCH_LGCORLR" length="12" type="real" minval="-179.9999999" maxval="179.9999999" unit="degrees"/>
+            <field name="CG_PATCH_LTCORLL" length="11" type="real" minval="-89.9999999" maxval="89.9999999" unit="degrees"/>
+            <field name="CG_PATCH_LGCORLL" length="12" type="real" minval="-179.9999999" maxval="179.9999999" unit="degrees"/>
```
Comment (bradh) [inline core/src/main/resources/nitf_spec.xml]: `CG_PATCH_LNGCOLL`?

File: core/src/main/resources/nitf_spec.xml [hunk 9]
```diff
@@ -199,6 +199,214 @@
...
+            <field name="CG_PATCH_LAT_CONFIDENCE" length="9" type="real" minval="0.0" maxval="9.9999999" unit="degrees"/>
+            <field name="CG_PATCH_LONG_CONFIDENCE" length="9" type="real" minval="0.0" maxval="9.9999999" unit="degrees"/>
+        </if>
+
+        <if cond="CG_MAP_TYPE=MGRS">
+            <field name="CG_MGRS_CENT" length="23" type="string"/>
+            <field name="CG_MGRSCORUL" length="23" type="string"/>
+            <field name="CG_MGRSCORUR" length="23" type="string"/>
+            <field name="CG_MGRSCORLR" length="23" type="string"/>
+            <field name="CG_MGRSCORLL" length="23" type="string"/>
```
Comment (bradh) [inline core/src/main/resources/nitf_spec.xml]: `CG_MGRCORLL`?

File: core/src/main/resources/nitf_spec.xml [hunk 10]
```diff
@@ -199,6 +199,214 @@
...
+            <field name="CG_MGRS_CONFIDENCE" length="7" type="real" minval="0.0" maxval="9999.99" unit="meters"/>
+            <field name="CG_MGRS_PAD" length="11" type="string"/>
+        </if>
+
+        <if cond="CG_MAP_TYPE=NA  ">
+            <field name="CG_MAP_TYPE_BLANK" length="133" type="string"/>
+        </if>
+
+        <field name="CG_SPARE_A" length="144" type="string"/>
```
Comment (bradh) [inline core/src/main/resources/nitf_spec.xml]: Could use `fixed_value` to fill this if desired.

**Comment by derekwilhelm** [comment]
Checked the spec and updated the field names to match.

File: core/src/main/resources/nitf_spec.xml (hunk 9 above)
Comment (derekwilhelm) [reply core/src/main/resources/nitf_spec.xml]: Unfortunately, yes, the inconsistency is not ideal.  However, updated to match the spec.

File: core/src/main/resources/nitf_spec.xml (hunk 8 above)
Comment (derekwilhelm) [reply core/src/main/resources/nitf_spec.xml]: Unfortunately, yes, the inconsistency is not ideal.  However, updated to match the spec.

File: core/src/main/resources/nitf_spec.xml [hunk 11]
```diff
@@ -199,6 +199,214 @@
...
+        <field name="T_HHMMSSUTC" length="6" type="string"/>
+        <field name="T_HHMMSSLOCAL" length="6" type="string"/>
+        <field name="CG_SRAC" length="11" type="real" minval="0.0" maxval="99999999.99" unit="meters"/>
+        <field name="CG_SLANT_CONFIDENCE" length="7" type="real" minval="0.0" maxval="9999.99" unit="meters"/>
+        <field name="CG_CROSS" length="11" type="real" minval="0.0" maxval="99999999.99" unit="meters"/>
+        <field name="CG_CROSS_CONFIDENCE" length="7" type="real" minval="0.0" maxval="9999.99" unit="meters"/>
+        <field name="CG_CAAC" length="9" type="real" minval="-179.0" maxval="179.0" unit="degrees"/>
+        <field name="CG_CONE_CONFIDENCE" length="6" type="real" minval="0.0" maxval="0.9999" unit="degrees"/>
+        <field name="CG_GPSAC" length="8" type="real" minval="-89.0" maxval="89.0" unit="degrees"/>
+        <field name="CG_SQUINT_CONFIDENCE" length="6" type="real" minval="0.0" maxval="0.9999" unit="degrees"/>
```
[the remark this answers is not in the document]
Comment (derekwilhelm) [reply core/src/main/resources/nitf_spec.xml]: Yes, updated to match the spec

File: core/src/main/resources/nitf_spec.xml [hunk 12]
```diff
@@ -199,6 +199,214 @@
...
+        <field name="RD_CEN_FREQ" length="4" type="string"/>
+        <field name="RD_MODE" length="5" type="string"/>
+        <field name="RD_PATCH_NO" length="4" type="integer" minval="0" maxval="9999"/>
+        <field name="CMPLX_DOMAIN" length="5" type="string"/>
+        <field name="CMPLX_MAG_SCALE_TYPE" length="4" type="string"/>
+        <field name="CMPLX_LIN_SCALE" length="7" type="real" minval="0.000001" maxval="99999.9"/>
+        <field name="CMPLX_AVG_POWER" length="7" type="real" minval="0" maxval="99999.9"/>
+        <field name="CMPLX_LINLOG_TP" length="5" type="integer" minval="0" maxval="65535"/>
+        <field name="CMPLX_PHASE_SCALING_TYPE" length="3" type="string"/>
+        <field name="CMPLX_PHASE_SCALE" length="2" type="integer" minval="0" maxval="32"/>
```
[the remark this answers is not in the document]
Comment (derekwilhelm) [reply core/src/main/resources/nitf_spec.xml]: Yes, updated to match the spec

File: core/src/main/resources/nitf_spec.xml [hunk 13]
```diff
@@ -199,6 +199,214 @@
...
+    <tre name="CMETAA" length="1572" location="image">
+        <field name="RELATED_TRES" length="2" type="integer" minval="0" maxval="20"/>
+        <field name="ADDITIONAL_TRES" length="120" type="string"/>
+        <field name="RD_PRC_NO" length="12" type="string"/>
+        <field name="IF_PROCESS" length="4" type="string"/>
+        <field name="RD_CEN_FREQ" length="4" type="string"/>
+        <field name="RD_MODE" length="5" type="string"/>
+        <field name="RD_PATCH_NO" length="4" type="integer" minval="0" maxval="9999"/>
+        <field name="CMPLX_DOMAIN" length="5" type="string"/>
+        <field name="CMPLX_MAG_SCALE_TYPE" length="4" type="string"/>
```
[the remark this answers is not in the document]
Comment (derekwilhelm) [reply core/src/main/resources/nitf_spec.xml]: Updated to match the spec

**Pull request merged**